wkcuber/wkcuber/export_wkw_as_nifti.py

```diff
@@ -0,0 +1,201 @@
+        "(By default, data for the full bounding box of the dataset is generated)",
+        default=None,
+        type=parse_bounding_box,
+    )
+
+    parser.add_argument(
+        "--mag", "-m", help="The magnification that should be read", default=1
+    )
+
+    parser.add_argument(
+        "--downsample", help="Downsample each nifti image", default=1, type=int
+    )
+
+    parser.add_argument(
+        "--padded_bbox",
+        help="After reading from --source_bbox in the input file, the data is zero-padded to fit --padded_bbox. --source_bbox must be contained in --padded_bbox. Format is: x,y,z,width,height,depth"
+        "x,y,z is the offset of the wkw layer into the final bounding box; "
+        "width,height,depth corresponds to final dimensions",
+        default=None,
+        type=parse_bounding_box,
+    )
+
+    parser.add_argument(
+        "--cropping_bbox",
+        help="After reading (and potentially padding) the input data, the output is cropped according to --cropping_bbox. --cropping_bbox must be contained in --padded_bbox"
+        "Format is: x,y,z,width,height,depth",
+        default=None,
+        type=parse_bounding_box,
+    )
+
+    add_verbose_flag(parser)
+    add_distribution_flags(parser)
+
+    return parser
+
+
+def export_layer_to_nifti(
+    wkw_file_path: Path,
+    source_bbox: BoundingBox,
+    mag: Mag,
+    layer_name: str,
+    destination_path: Path,
+    name: str,
+    padded_bbox: Optional[BoundingBox] = None,
+    cropping_bbox: Optional[BoundingBox] = None,
+) -> None:
+    dataset = Dataset(wkw_file_path)
+    layer = dataset.get_layer(layer_name)
+    mag_layer = layer.get_mag(mag)
+
+    is_segmentation_layer = layer.category == LayerCategories.SEGMENTATION_TYPE
+
+    data = mag_layer.read(source_bbox.topleft, source_bbox.size)
+    data = data.transpose(1, 2, 3, 0)
+    logging.info(f"Shape with layer {data.shape}")
+
+    data = np.array(data)
+    if is_segmentation_layer and data.max() > 0:
+        factor = np.iinfo("uint8").max / data.max()
+        data = data * factor
+        data = data.astype(np.dtype("uint8"))
+
+    if padded_bbox:
+        data_bbox_relative_to_pad_bbox = BoundingBox(
+            padded_bbox.topleft, data.shape[:3]
+        )
+        pad_bbox_as_origin = BoundingBox((0, 0, 0), padded_bbox.size)
+
+        assert pad_bbox_as_origin.contains_bbox(
+            data_bbox_relative_to_pad_bbox
+        ), "padded_bbox should contain source_bbox"
+
+        padding_per_axis = []
+
+        for i in range(3):
+            if padded_bbox.size[i] == data.shape[i]:
+                padding_per_axis.append((0, 0))
+            else:
+                left_pad = data_bbox_relative_to_pad_bbox.topleft[i]
+                right_pad = (
+                    pad_bbox_as_origin.size[i]
+                    - data_bbox_relative_to_pad_bbox.bottomright[i]
+                )
+                padding_per_axis.append((left_pad, right_pad))
```

Comment:
To be honest, I'm a bit lost with the semantics of the `--padded_bbox` parameter :see_no_evil: I thought, the padding would simply be the difference between `--source_bbox` and the `--padded_bbox`. My mental model is as follows:

![image](https://user-images.githubusercontent.com/2486553/133736127-1f4511cd-3920-4aea-8899-834139f4f475.png)


However, your code does something different and I can't really figure out what the semantics are. The topleft of the `padded_bbox` seems to be simply used for left padding which could also be a way of designing the parameter it, but (1) then I would simply call the parameter `--padding` and make its format `(leftpadx, leftpady, leftpadz, rightpadx, rightpady, rightpadz)` and (2) right_pad doesn't fit to that semantics.

Maybe I'm missing something here, but I think it's easier to talk about a concrete example:

```
- source_bbox goes from 25, 25, 25   to   55, 55, 55
- padded_bbox goes from 10, 10, 10   to   70, 70, 70

- then, leftpad would currently be 10, but I'd expect 15
- rightpad would be 60 - (55 - (25 - 10)) = 20, but I'd expect 15
```

Reply:
Yes you are right, we need to get that right. I also stumbled over the semantics and what the best approach would be. I think I designed this interface in march to just throw in the bounding boxes that I used to transform the nifti to wkw. Simply using `--padding` would probably make this interface much easier. When I need to use it again later, I just add the necessary logic on the client side to calculate the necessary padding, this will also be easier if the semantics here are crystal clear.